### ISSUE TITLE
How to get the favicon of <webview> tag?

### ISSUE THREAD

**Issue by DemoHn** [atom/electron#1079]
As the title mentioned, when I was trying to emulate a browser using the `<webview>` tag, I hope to get the favicon as well as the title of the webpage,like what actual browsers do.
However, I only find `<webview>.getTitle()` method to accomplish the task where there is no method like `<webview>.getFavicon()`, so maybe there is some other ways to get the favicon?

**Comment by frankhale** [comment]
:+1:

**Comment by mark-hahn** [comment]
Use `"http://www.google.com/s2/favicons?domain=#{domain}"`

**Comment by DemoHn** [comment]
Thank you for your link @mark-hahn .^_^
But I don't think it is a perfect way to rely on external resources.Because some websites (including my school's internal forum) cannot be captured by Google.

_And in fact, Google is totally blocked in my country (WTF)_

I have a solution about this, but I think it is really weird:

1. in `index.html`
```
<webview src='https://github.com' preload="./__insert.js"></webview>
```
2. And in `__insert.js`
```
var ipc = require('ipc'),
    url = require('url');

window.onload = function () {
    ipc.sendToHost(getFaviconLink(),'webview:favicon');
}

function getFaviconLink(){
    var href = window.location.href;

    var final_favicon = "";
    var elem = document.getElementsByTagName('link');
    for(var i=0;i<elem.length;i++){
        if(/icon/.test(elem[i].rel) == true){
            final_favicon = elem[i].href;

            // if rel="shortcut icon" ,it should be considered first.(WTF!!)
            if(/^shortcut /.test(elem[i].rel) == true){
                break;
            }
        }
    }

    if(final_favicon == ""){
        return null;
    }else{
        return url.resolve(href,final_favicon);
    }
}
```
Just use `ipc.sendToHost` to get the whole content of the guest page where favicon is included and send it to the main frame.

**Comment by mark-hahn** [comment]
>  get the whole content of the guest page where favicon is included

Web pages don't include favicons.  You have to do an extra read from the
server.  Webview never loads the favicon, just the page.  To do it like a
browser does you will have to do that separate read.

**Comment by zcbenz** [comment]
Chromium has [an API](https://code.google.com/p/chromium/codesearch#chromium/src/content/public/browser/web_contents_observer.h&sq=package:chromium&l=310) to return the favicon urls of the page, as @mark-hahn said you still have to download the icons separately though.

**Comment by DemoHn** [comment]
I'm sorry for not describing it clearly..I mean I just fetch the favicon url from the guest page and download the icon separately.
And if I want to use the API, is `webview.addEventListener('did-update-favicon-url')` reliable? Or still wating for upgrate?

**Comment by mark-hahn** [comment]
I didn't know about `did-update-favicon-url`.  It isn't in https://github.com/atom/atom-shell/blob/master/docs/api/web-view-tag.md.  I'll have to read the source.

**Comment by DemoHn** [comment]
Yeah, that's just my guess.There is no such `did-update-favicon-url` listener.=w=

**Comment by mark-hahn** [comment]
> I mean I just fetch the favicon url from the guest page

There isn't a URL in the page.  There is nothing about them in the page at
all.  The browser has to look around in the server to see if it can find
anything.  There are standard files to look for.  I know about the old IE
favicon file and the newer apple one but there are probably even more
possibilities.

This is why google started offering the free favicon service I mentioned.
It works well for me and most others in this situation.

**Comment by DemoHn** [comment]
Sorry but I don't agree with you.
There is a URL in the page,just written like this : `<link rel="icon" href="{favicon link}">`

For example, you can see such code:`<link rel="icon" type="image/x-icon" href="https://assets-cdn.github.com/favicon.ico">` when viewing the source code of https://github.com/atom/atom-shell/issues/1079. That's GitHub's favicon URL.

Or Stackoverflow: `<link rel="shortcut icon" href="//cdn.sstatic.net/stackoverflow/img/favicon.ico?v=038622610830">`

Sometimes there may not have a URL in the page, as you mentioned. Browsers may handle it by requesting `/favicon.ico` . like: [baidu](http://www.baidu.com). This page does not include the favicon url in the source code, so the default one `/favicon.ico` is available.

**Comment by mark-hahn** [comment]
Thanks.  That is news to me.

So having the url in the page speeds things up a bit. It removes the
possible need to search several places.  Not much speedup though because
you still have to load a file either way.  And your code has to be prepared
for either case.  And the server still needs to host in the old place for
old browsers.

It is great that you can use a better file name and location.  The old IE
standard always bothered me.

**Comment by zcbenz** [comment]
Fixed by #1438
